File: src/main/java/io/nshusa/rsam/graphics/render/Rasterizer2D.java
package io.nshusa.rsam.graphics.render;

public class Rasterizer2D {
    public static float[] depthBuffer;
    public static int[] pixels;
    public static int width;
    public static int height;
    public static int topY;
    public static int bottomY;
    public static int leftX;
    public static int bottomX;
    public static int lastX;
    public static int viewportCenterX;
    public static int viewportCenterY;

    public Rasterizer2D() {
    }

    public static void initDrawingArea(int height, int width, int[] pixels, float[] depth) {
        depthBuffer = depth;
        pixels = pixels;
        width = width;
        height = height;
        setDrawingArea(height, 0, width, 0);
    }

    public static void drawTransparentGradientBox(int leftX, int topY, int width, int height, int topColour, int bottomColour, int opacity) {
        int gradientProgress = 0;
        int progressPerPixel = 65536 / height;
        if (leftX < leftX) {
            width -= leftX - leftX;
            leftX = leftX;
        }

        if (topY < topY) {
            gradientProgress += (topY - topY) * progressPerPixel;
            height -= topY - topY;
            topY = topY;
        }

        if (leftX + width > bottomX) {
            width = bottomX - leftX;
        }

        if (topY + height > bottomY) {
            height = bottomY - topY;
        }

        int leftOver = width - width;
        int transparency = 256 - opacity;
        int pixelIndex = leftX + topY * width;

        for(int rowIndex = 0; rowIndex < height; ++rowIndex) {
            int gradient = 65536 - gradientProgress >> 8;
            int inverseGradient = gradientProgress >> 8;
            int gradientColour = ((topColour & 16711935) * gradient + (bottomColour & 16711935) * inverseGradient & -16711936) + ((topColour & '\uff00') * gradient + (bottomColour & '\uff00') * inverseGradient & 16711680) >>> 8;
            int transparentPixel = ((gradientColour & 16711935) * opacity >> 8 & 16711935) + ((gradientColour & '\uff00') * opacity >> 8 & '\uff00');

            for(int columnIndex = 0; columnIndex < width; ++columnIndex) {
                int backgroundPixel = pixels[pixelIndex];
                backgroundPixel = ((backgroundPixel & 16711935) * transparency >> 8 & 16711935) + ((backgroundPixel & '\uff00') * transparency >> 8 & '\uff00');
                pixels[pixelIndex++] = transparentPixel + backgroundPixel;
            }

            pixelIndex += leftOver;
            gradientProgress += progressPerPixel;
        }

    }

    public static void defaultDrawingAreaSize() {
        leftX = 0;
        topY = 0;
        bottomX = width;
        bottomY = height;
        lastX = bottomX;
        viewportCenterX = bottomX / 2;
    }

    public static void setDrawingArea(int bottomY, int leftX, int rightX, int topY) {
        if (leftX < 0) {
            leftX = 0;
        }

        if (topY < 0) {
            topY = 0;
        }

        if (rightX > width) {
            rightX = width;
        }

        if (bottomY > height) {
            bottomY = height;
        }

        leftX = leftX;
        topY = topY;
        bottomX = rightX;
        bottomY = bottomY;
        lastX = bottomX;
        viewportCenterX = bottomX / 2;
        viewportCenterY = bottomY / 2;
    }

    public static void clear() {
        int i = width * height;

        for(int j = 0; j < i; ++j) {
            pixels[j] = 0;
            depthBuffer[j] = 3.4028235E38F;
        }

    }

    public static void drawBox(int leftX, int topY, int width, int height, int rgbColour) {
        if (leftX < leftX) {
            width -= leftX - leftX;
            leftX = leftX;
        }

        if (topY < topY) {
            height -= topY - topY;
            topY = topY;
        }

        if (leftX + width > bottomX) {
            width = bottomX - leftX;
        }

        if (topY + height > bottomY) {
            height = bottomY - topY;
        }

        int leftOver = width - width;
        int pixelIndex = leftX + topY * width;

        for(int rowIndex = 0; rowIndex < height; ++rowIndex) {
            for(int columnIndex = 0; columnIndex < width; ++columnIndex) {
                pixels[pixelIndex++] = rgbColour;
            }

            pixelIndex += leftOver;
        }

    }

    public static void drawTransparentBox(int leftX, int topY, int width, int height, int rgbColour, int opacity) {
        if (leftX < leftX) {
            width -= leftX - leftX;
            leftX = leftX;
        }

        if (topY < topY) {
            height -= topY - topY;
            topY = topY;
        }

        if (leftX + width > bottomX) {
            width = bottomX - leftX;
        }

        if (topY + height > bottomY) {
            height = bottomY - topY;
        }

        int transparency = 256 - opacity;
        int red = (rgbColour >> 16 & 255) * opacity;
        int green = (rgbColour >> 8 & 255) * opacity;
        int blue = (rgbColour & 255) * opacity;
        int leftOver = width - width;
        int pixelIndex = leftX + topY * width;

        for(int rowIndex = 0; rowIndex < height; ++rowIndex) {
            for(int columnIndex = 0; columnIndex < width; ++columnIndex) {
                int otherRed = (pixels[pixelIndex] >> 16 & 255) * transparency;
                int otherGreen = (pixels[pixelIndex] >> 8 & 255) * transparency;
                int otherBlue = (pixels[pixelIndex] & 255) * transparency;
                int transparentColour = (red + otherRed >> 8 << 16) + (green + otherGreen >> 8 << 8) + (blue + otherBlue >> 8);
                pixels[pixelIndex++] = transparentColour;
            }

            pixelIndex += leftOver;
        }

    }

    public static void drawBoxOutline(int leftX, int topY, int width, int height, int rgbColour) {
        drawHorizontalLine(leftX, topY, width, rgbColour);
        drawHorizontalLine(leftX, topY + height - 1, width, rgbColour);
        drawVerticalLine(leftX, topY, height, rgbColour);
        drawVerticalLine(leftX + width - 1, topY, height, rgbColour);
    }

    public static void drawHorizontalLine(int xPosition, int yPosition, int width, int rgbColour) {
        if (yPosition >= topY && yPosition < bottomY) {
            if (xPosition < leftX) {
                width -= leftX - xPosition;
                xPosition = leftX;
            }

            if (xPosition + width > bottomX) {
                width = bottomX - xPosition;
            }

            int pixelIndex = xPosition + yPosition * width;

            for(int i = 0; i < width; ++i) {
                pixels[pixelIndex + i] = rgbColour;
            }

        }
    }

    public static void drawVerticalLine(int xPosition, int yPosition, int height, int rgbColour) {
        if (xPosition >= leftX && xPosition < bottomX) {
            if (yPosition < topY) {
                height -= topY - yPosition;
                yPosition = topY;
            }

            if (yPosition + height > bottomY) {
                height = bottomY - yPosition;
            }

            int pixelIndex = xPosition + yPosition * width;

            for(int rowIndex = 0; rowIndex < height; ++rowIndex) {
                pixels[pixelIndex + rowIndex * width] = rgbColour;
            }

        }
    }

    public static void drawTransparentBoxOutline(int leftX, int topY, int width, int height, int rgbColour, int opacity) {
        drawTransparentHorizontalLine(leftX, topY, width, rgbColour, opacity);
        drawTransparentHorizontalLine(leftX, topY + height - 1, width, rgbColour, opacity);
        if (height >= 3) {
            drawTransparentVerticalLine(leftX, topY + 1, height - 2, rgbColour, opacity);
            drawTransparentVerticalLine(leftX + width - 1, topY + 1, height - 2, rgbColour, opacity);
        }

    }

    public static void drawTransparentHorizontalLine(int xPosition, int yPosition, int width, int rgbColour, int opacity) {
        if (yPosition >= topY && yPosition < bottomY) {
            if (xPosition < leftX) {
                width -= leftX - xPosition;
                xPosition = leftX;
            }

            if (xPosition + width > bottomX) {
                width = bottomX - xPosition;
            }

            int transparency = 256 - opacity;
            int red = (rgbColour >> 16 & 255) * opacity;
            int green = (rgbColour >> 8 & 255) * opacity;
            int blue = (rgbColour & 255) * opacity;
            int pixelIndex = xPosition + yPosition * width;

            for(int i = 0; i < width; ++i) {
                int otherRed = (pixels[pixelIndex] >> 16 & 255) * transparency;
                int otherGreen = (pixels[pixelIndex] >> 8 & 255) * transparency;
                int otherBlue = (pixels[pixelIndex] & 255) * transparency;
                int transparentColour = (red + otherRed >> 8 << 16) + (green + otherGreen >> 8 << 8) + (blue + otherBlue >> 8);
                pixels[pixelIndex++] = transparentColour;
            }

        }
    }

    public static void drawTransparentVerticalLine(int xPosition, int yPosition, int height, int rgbColour, int opacity) {
        if (xPosition >= leftX && xPosition < bottomX) {
            if (yPosition < topY) {
                height -= topY - yPosition;
                yPosition = topY;
            }

            if (yPosition + height > bottomY) {
                height = bottomY - yPosition;
            }

            int transparency = 256 - opacity;
            int red = (rgbColour >> 16 & 255) * opacity;
            int green = (rgbColour >> 8 & 255) * opacity;
            int blue = (rgbColour & 255) * opacity;
            int pixelIndex = xPosition + yPosition * width;

            for(int i = 0; i < height; ++i) {
                int otherRed = (pixels[pixelIndex] >> 16 & 255) * transparency;
                int otherGreen = (pixels[pixelIndex] >> 8 & 255) * transparency;
                int otherBlue = (pixels[pixelIndex] & 255) * transparency;
                int transparentColour = (red + otherRed >> 8 << 16) + (green + otherGreen >> 8 << 8) + (blue + otherBlue >> 8);
                pixels[pixelIndex] = transparentColour;
                pixelIndex += width;
            }

        }
    }
}

File: src/main/kotlin/io/nshusa/rsam/graphics/render/Rasterizer2D.kt
package io.nshusa.rsam.graphics.render

open class Rasterizer2D {
    companion object {
        lateinit var depthBuffer: FloatArray
        var pixels: IntArray? = null
        var width: Int = 0
        var height: Int = 0
        var topY: Int = 0
        var bottomY: Int = 0
        var leftX: Int = 0
        var bottomX: Int = 0
        var lastX: Int = 0
        var viewportCenterX: Int = 0
        var viewportCenterY: Int = 0

        fun initDrawingArea(height: Int, width: Int, pixels: IntArray, depth: FloatArray) {
            var height = height
            var width = width
            var pixels = pixels
            depthBuffer = depth
            pixels = pixels
            width = width
            height = height
            setDrawingArea(height, 0, width, 0)
        }

        fun drawTransparentGradientBox(leftX: Int, topY: Int, width: Int, height: Int, topColour: Int, bottomColour: Int, opacity: Int) {
            var leftX = leftX
            var topY = topY
            var width = width
            var height = height
            var gradientProgress = 0
            val progressPerPixel = 65536 / height
            if (leftX < leftX) {
                width -= leftX - leftX
                leftX = leftX
            }

            if (topY < topY) {
                gradientProgress += (topY - topY) * progressPerPixel
                height -= topY - topY
                topY = topY
            }

            if (leftX + width > bottomX) {
                width = bottomX - leftX
            }

            if (topY + height > bottomY) {
                height = bottomY - topY
            }

            val leftOver = width - width
            val transparency = 256 - opacity
            var pixelIndex = leftX + topY * width

            for (rowIndex in 0 until height) {
                val gradient = 65536 - gradientProgress shr 8
                val inverseGradient = gradientProgress shr 8
                val gradientColour = (((topColour and 16711935) * gradient + (bottomColour and 16711935) * inverseGradient and -16711936) + ((topColour and '\uff00'.toInt()) * gradient + (bottomColour and '\uff00'.toInt()) * inverseGradient and 16711680)).ushr(8)
                val transparentPixel = ((gradientColour and 16711935) * opacity shr 8 and 16711935) + ((gradientColour and '\uff00'.toInt()) * opacity shr 8 and '\uff00'.toInt())

                for (columnIndex in 0 until width) {
                    var backgroundPixel = pixels!![pixelIndex]
                    backgroundPixel = ((backgroundPixel and 16711935) * transparency shr 8 and 16711935) + ((backgroundPixel and '\uff00'.toInt()) * transparency shr 8 and '\uff00'.toInt())
                    pixels!![pixelIndex++] = transparentPixel + backgroundPixel
                }

                pixelIndex += leftOver
                gradientProgress += progressPerPixel
            }

        }

        fun defaultDrawingAreaSize() {
            leftX = 0
            topY = 0
            bottomX = width
            bottomY = height
            lastX = bottomX
            viewportCenterX = bottomX / 2
        }

        fun setDrawingArea(bottomY: Int, leftX: Int, rightX: Int, topY: Int) {
            var bottomY = bottomY
            var leftX = leftX
            var rightX = rightX
            var topY = topY
            if (leftX < 0) {
                leftX = 0
            }

            if (topY < 0) {
                topY = 0
            }

            if (rightX > width) {
                rightX = width
            }

            if (bottomY > height) {
                bottomY = height
            }

            leftX = leftX
            topY = topY
            bottomX = rightX
            bottomY = bottomY
            lastX = bottomX
            viewportCenterX = bottomX / 2
            viewportCenterY = bottomY / 2
        }

        fun clear() {
            val i = width * height

            for (j in 0 until i) {
                pixels!![j] = 0
                depthBuffer[j] = 3.4028235E38f
            }

        }

        fun drawBox(leftX: Int, topY: Int, width: Int, height: Int, rgbColour: Int) {
            var leftX = leftX
            var topY = topY
            var width = width
            var height = height
            if (leftX < leftX) {
                width -= leftX - leftX
                leftX = leftX
            }

            if (topY < topY) {
                height -= topY - topY
                topY = topY
            }

            if (leftX + width > bottomX) {
                width = bottomX - leftX
            }

            if (topY + height > bottomY) {
                height = bottomY - topY
            }

            val leftOver = width - width
            var pixelIndex = leftX + topY * width

            for (rowIndex in 0 until height) {
                for (columnIndex in 0 until width) {
                    pixels!![pixelIndex++] = rgbColour
                }

                pixelIndex += leftOver
            }

        }

        fun drawTransparentBox(leftX: Int, topY: Int, width: Int, height: Int, rgbColour: Int, opacity: Int) {
            var leftX = leftX
            var topY = topY
            var width = width
            var height = height
            if (leftX < leftX) {
                width -= leftX - leftX
                leftX = leftX
            }

            if (topY < topY) {
                height -= topY - topY
                topY = topY
            }

            if (leftX + width > bottomX) {
                width = bottomX - leftX
            }

            if (topY + height > bottomY) {
                height = bottomY - topY
            }

            val transparency = 256 - opacity
            val red = (rgbColour shr 16 and 255) * opacity
            val green = (rgbColour shr 8 and 255) * opacity
            val blue = (rgbColour and 255) * opacity
            val leftOver = width - width
            var pixelIndex = leftX + topY * width

            for (rowIndex in 0 until height) {
                for (columnIndex in 0 until width) {
                    val otherRed = (pixels!![pixelIndex] shr 16 and 255) * transparency
                    val otherGreen = (pixels!![pixelIndex] shr 8 and 255) * transparency
                    val otherBlue = (pixels!![pixelIndex] and 255) * transparency
                    val transparentColour = (red + otherRed shr 8 shl 16) + (green + otherGreen shr 8 shl 8) + (blue + otherBlue shr 8)
                    pixels!![pixelIndex++] = transparentColour
                }

                pixelIndex += leftOver
            }

        }

        fun drawBoxOutline(leftX: Int, topY: Int, width: Int, height: Int, rgbColour: Int) {
            drawHorizontalLine(leftX, topY, width, rgbColour)
            drawHorizontalLine(leftX, topY + height - 1, width, rgbColour)
            drawVerticalLine(leftX, topY, height, rgbColour)
            drawVerticalLine(leftX + width - 1, topY, height, rgbColour)
        }

        fun drawHorizontalLine(xPosition: Int, yPosition: Int, width: Int, rgbColour: Int) {
            var xPosition = xPosition
            var width = width
            if (yPosition >= topY && yPosition < bottomY) {
                if (xPosition < leftX) {
                    width -= leftX - xPosition
                    xPosition = leftX
                }

                if (xPosition + width > bottomX) {
                    width = bottomX - xPosition
                }

                val pixelIndex = xPosition + yPosition * width

                for (i in 0 until width) {
                    pixels!![pixelIndex + i] = rgbColour
                }

            }
        }

        fun drawVerticalLine(xPosition: Int, yPosition: Int, height: Int, rgbColour: Int) {
            var yPosition = yPosition
            var height = height
            if (xPosition >= leftX && xPosition < bottomX) {
                if (yPosition < topY) {
                    height -= topY - yPosition
                    yPosition = topY
                }

                if (yPosition + height > bottomY) {
                    height = bottomY - yPosition
                }

                val pixelIndex = xPosition + yPosition * width

                for (rowIndex in 0 until height) {
                    pixels!![pixelIndex + rowIndex * width] = rgbColour
                }

            }
        }

        fun drawTransparentBoxOutline(leftX: Int, topY: Int, width: Int, height: Int, rgbColour: Int, opacity: Int) {
            drawTransparentHorizontalLine(leftX, topY, width, rgbColour, opacity)
            drawTransparentHorizontalLine(leftX, topY + height - 1, width, rgbColour, opacity)
            if (height >= 3) {
                drawTransparentVerticalLine(leftX, topY + 1, height - 2, rgbColour, opacity)
                drawTransparentVerticalLine(leftX + width - 1, topY + 1, height - 2, rgbColour, opacity)
            }

        }

        fun drawTransparentHorizontalLine(xPosition: Int, yPosition: Int, width: Int, rgbColour: Int, opacity: Int) {
            var xPosition = xPosition
            var width = width
            if (yPosition >= topY && yPosition < bottomY) {
                if (xPosition < leftX) {
                    width -= leftX - xPosition
                    xPosition = leftX
                }

                if (xPosition + width > bottomX) {
                    width = bottomX - xPosition
                }

                val transparency = 256 - opacity
                val red = (rgbColour shr 16 and 255) * opacity
                val green = (rgbColour shr 8 and 255) * opacity
                val blue = (rgbColour and 255) * opacity
                var pixelIndex = xPosition + yPosition * width

                for (i in 0 until width) {
                    val otherRed = (pixels!![pixelIndex] shr 16 and 255) * transparency
                    val otherGreen = (pixels!![pixelIndex] shr 8 and 255) * transparency
                    val otherBlue = (pixels!![pixelIndex] and 255) * transparency
                    val transparentColour = (red + otherRed shr 8 shl 16) + (green + otherGreen shr 8 shl 8) + (blue + otherBlue shr 8)
                    pixels!![pixelIndex++] = transparentColour
                }

            }
        }

        fun drawTransparentVerticalLine(xPosition: Int, yPosition: Int, height: Int, rgbColour: Int, opacity: Int) {
            var yPosition = yPosition
            var height = height
            if (xPosition >= leftX && xPosition < bottomX) {
                if (yPosition < topY) {
                    height -= topY - yPosition
                    yPosition = topY
                }

                if (yPosition + height > bottomY) {
                    height = bottomY - yPosition
                }

                val transparency = 256 - opacity
                val red = (rgbColour shr 16 and 255) * opacity
                val green = (rgbColour shr 8 and 255) * opacity
                val blue = (rgbColour and 255) * opacity
                var pixelIndex = xPosition + yPosition * width

                for (i in 0 until height) {
                    val otherRed = (pixels!![pixelIndex] shr 16 and 255) * transparency
                    val otherGreen = (pixels!![pixelIndex] shr 8 and 255) * transparency
                    val otherBlue = (pixels!![pixelIndex] and 255) * transparency
                    val transparentColour = (red + otherRed shr 8 shl 16) + (green + otherGreen shr 8 shl 8) + (blue + otherBlue shr 8)
                    pixels!![pixelIndex] = transparentColour
                    pixelIndex += width
                }

            }
        }
    }
}